File: Week_01/cpp/src/_2_AddTwoNumbers.cpp
#include <vector>
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
class Solution
{
public:
    // 逐位计算并考虑进位值
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode head(0), *p = &head;
        int add = 0, a;
        while (l1 || l2 || add)
        {
            a = 0 + add;
            if (l1)
            {
                a += l1->val;
                l1 = l1->next;
            }
            if (l2)
            {
                a += l2->val;
                l2 = l2->next;
            }
            p->next = new ListNode(a % 10);
            p = p->next;
            add = a / 10;
        }
        return head.next;
    }
};

File: Week_01/kotlin/src/_2_AddTwoNumbers.kt
class _2_ {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var p1 = l1
        var p2 = l2
        var carry = 0
        val dummy = ListNode(0)
        var p = dummy
        while (p1 != null || p2 != null || carry != 0) {
            var sum = carry
            if (p1 != null) {
                sum += p1.`val`
                p1 = p1.next
            }
            if (p2 != null) {
                sum += p2.`val`
                p2 = p2.next
            }
            p.next = ListNode(sum % 10)
            p = p.next!!
            carry = sum / 10
        }
        return dummy.next
    }
}